model/project.go

```diff
@@ -564,42 +564,98 @@ var (
-	expansions.Put("workdir", d.WorkDir)
 	expansions.Put("revision", t.Revision)
 	expansions.Put("project", t.Project)
+
+	expansions.Put("workdir", h.Distro.WorkDir)
+	expansions.Put("distro_id", h.Distro.Id)
+
+	v, err := version.FindOneId(t.Version)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding version")
+	}
 	expansions.Put("branch_name", v.Branch)
 	expansions.Put("author", v.Author)
-	expansions.Put("distro_id", d.Id)
 	expansions.Put("created_at", v.CreateTime.Format(build.IdTimeLayout))
 
 	if evergreen.IsPatchRequester(v.Requester) {
 		expansions.Put("is_patch", "true")
 		expansions.Put("revision_order_id", fmt.Sprintf("%s_%d", v.Author, v.RevisionOrderNumber))
+		p, err := patch.FindOne(patch.ByVersion(t.Version))
+		if err != nil {
+			return nil, errors.Wrap(err, "error finding patch")
+		}
 
 		if v.Requester == evergreen.GithubPRRequester && p != nil {
 			expansions.Put("github_pr_number", fmt.Sprintf("%d", p.GithubPatchData.PRNumber))
 			expansions.Put("github_org", p.GithubPatchData.BaseOwner)
 			expansions.Put("github_repo", p.GithubPatchData.BaseRepo)
 			expansions.Put("github_author", p.GithubPatchData.Author)
 		}
-
 	} else {
 		expansions.Put("revision_order_id", strconv.Itoa(v.RevisionOrderNumber))
 	}
 
-	for _, e := range d.Expansions {
+	if t.TriggerID != "" {
+		expansions.Put("trigger_event_identifier", t.TriggerID)
+		expansions.Put("trigger_event_type", t.TriggerType)
+		expansions.Put("trigger_id", t.TriggerEvent)
+		var upstreamProjectID string
+		if t.TriggerType == ProjectTriggerLevelTask {
+			upstreamTask, err := task.FindOneId(t.TriggerID)
+			if err != nil {
+				return nil, errors.Wrap(err, "error finding task")
+			}
+			expansions.Put("trigger_status", upstreamTask.Status)
+			expansions.Put("trigger_revision", upstreamTask.Revision)
+			upstreamProjectID = upstreamTask.Project
+		} else if t.TriggerType == ProjectTriggerLevelBuild {
+			upstreamBuild, err := build.FindOneId(t.TriggerID)
+			if err != nil {
+				return nil, errors.Wrap(err, "error finding build")
+			}
+			expansions.Put("trigger_status", upstreamBuild.Status)
+			expansions.Put("trigger_revision", upstreamBuild.Revision)
+			upstreamProjectID = upstreamBuild.Project
+		}
+		upstreamProject, err := FindOneProjectRef(upstreamProjectID)
+		if err != nil {
+			return nil, errors.Wrap(err, "error finding project")
+		}
+		if upstreamProject == nil {
+			return nil, errors.Errorf("upstream project %s not found", t.Project)
+		}
+		expansions.Put("trigger_repo_owner", upstreamProject.Owner)
+		expansions.Put("trigger_repo_name", upstreamProject.Repo)
+		expansions.Put("trigger_branch", upstreamProject.Branch)
+	}
+
+	for _, e := range h.Distro.Expansions {
 		expansions.Put(e.Key, e.Value)
 	}
+	proj := &Project{}
+	err = LoadProjectInto([]byte(v.Config), t.Project, proj)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling project")
+	}
+	bv := proj.FindBuildVariant(t.BuildVariant)
```

Comment:
Does this just move parsing the project from the server to the agent?

Reply:
this function has to run on the server as it does database calls.

Reply:
We're doing it in both places now (here only out of necessity to get the variant config)